packages/flutter/lib/src/material/search.dart

```diff
@@ -0,0 +1,442 @@
+  ///
+  ///  * [AppBar.actions], the intended use for the return value of this method.
+  List<Widget> buildActions(BuildContext context);
+
+  /// The theme used to style the [AppBar].
+  ///
+  /// By default, a white theme is used.
+  ///
+  /// See also:
+  ///
+  ///  * [AppBar.backgroundColor], which is set to [ThemeData.primaryColor].
+  ///  * [Appbar.iconTheme], which is set to [ThemeData.primaryIconTheme].
+  ///  * [AppBar.textTheme], which is set to [ThemeData.primaryTextTheme].
+  ///  * [AppBar.brightness], which is set to [ThemeData.primaryColorBrightness].
+  ThemeData appBarTheme(BuildContext context) {
+    assert(context != null);
+    final ThemeData theme = Theme.of(context);
+    assert(theme != null);
+    return theme.copyWith(
+      primaryColor: Colors.white,
+      primaryIconTheme: theme.primaryIconTheme.copyWith(color: Colors.grey),
+      primaryColorBrightness: Brightness.light,
+      primaryTextTheme: theme.textTheme,
+    );
+  }
+
+  /// The current query string shown in the [Appbar].
+  ///
+  /// The user manipulates this string via the keyboard.
+  ///
+  /// If the user taps on a suggestion provided by [buildSuggestions] this
+  /// string should be updated to that suggestion via the setter.
+  String get query => _queryTextController.text;
+  set query(String value) {
+    assert(query != null);
+    _queryTextController.text = value;
+  }
+
+  /// Transition from the suggestions returned by [buildSuggestions] to the
+  /// [query] results returned by [buildResults].
+  ///
+  /// If the user taps on a suggestion provided by [buildSuggestions] the
+  /// screen should typically transition to the page showing the search
+  /// results for the suggested query. This transition can be triggered
+  /// by calling this method.
+  ///
+  /// See also:
+  ///
+  ///  * [showSuggestions] to show the search suggestions again.
+  void showResults(BuildContext context) {
+    _focusNode.unfocus();
+    _currentBody = _SearchBody.results;
+  }
+
+  /// Transition from showing the results returned by [buildResults] to showing
+  /// the suggestions returned by [buildSuggestions].
+  ///
+  /// Calling this method will also put the input focus back into the search
+  /// field of the ApBar.
+  ///
+  /// If the results are currently shown this method can be used to go back
+  /// to showing the search suggestions.
+  ///
+  /// See also:
+  ///
+  ///  * [showResults] to show the search results.
+  void showSuggestions(BuildContext context) {
+    FocusScope.of(context).requestFocus(_focusNode);
+    _currentBody = _SearchBody.suggestions;
+  }
+
+  /// Closes the search page and returns to the underlying route.
+  ///
+  /// The value provided for `result` is used as the return value of the call
+  /// to [showSearch] that launched the search initially.
+  void close(BuildContext context, T result) {
+    _currentBody = null;
+    _focusNode.unfocus();
+    _result.complete(result);
+    Navigator.of(context)
```

Comment:
I think it's possible for the original route to have been removed while the search dialog is up, which would probably trigger a strange failure mode. If developers start running into that, we could make this code more defensive. In a future PR.

Reply:
Sounds good.